File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    answer = 0
    arr =[]
    def dfs(num, cnt):
        if visited[num-1]:
            arr.append(cnt)
            return
        
        visited[num-1] = True
        dfs(cards[num-1],cnt+1)
        
    
    visited=[False]*len(cards)
    
    for i in range(len(cards)):
        if not visited[i]:
            visited[i] = True
            dfs(cards[i],1)
    
    
    if len(arr)>=2:
        arr.sort()
        answer = arr[-1]*arr[-2]
    
    
    return answer

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;
class Solution {
    static int[] cards;
    public int solution(int[] cards) {
        this.cards = cards;
        int answer = 0;
        int n = cards.length;
        for(int i=0;i<n;i++){
            boolean[] visited = new boolean[n];

            int a = makeBox(cards[i], visited);
            for(int j=0; j<n;j++){
                if(!visited[j]){
                    int b = makeBox(cards[j],visited);
                    answer = Math.max(answer, a*b);
                }
            }
            
            
            
        }
        
        
        return answer;
    }
    public static int makeBox(int num, boolean[] visited){
        int cnt = 0;
        
        while(!visited[num-1]){
            visited[num-1] = true;
            num = cards[num-1];
            cnt++;
        }
        return cnt;
    }
}